test/support/annotations/matched_annotation.rb

```diff
@@ -0,0 +1,68 @@
+module Annotations
+  class MatchedAnnotation
+    attr_accessor :annotation, :requests, :method, :path, :description, :title,
+                  :ignore_duplicates, :starts_with, :ignored_count, :request_body
+
+    def initialize(hash)
+      self.requests = []
+      self.method = hash['method']
+      self.path = hash['path']
+      self.description = hash['description']
+      self.title = hash['title']
+      self.request_body = hash['request_body']
+      self.ignore_duplicates = hash.fetch('ignore_duplicates', false)
+      self.starts_with = hash['starts_with']
+      @hash = hash
+    end
+
+    def backend_service
+      if self.path.starts_with?('/pulp')
+        'pulp'
+      elsif self.path.starts_with?('/candlepin')
+        'candlepin'
+      else
+        'unknown'
+      end
+    end
+
+    def match_requests(vcr_requests)
+      found = []
+      vcr_requests.each do |request|
+        found << request if self.matches?(request)
+      end
+      found
+    end
+
+    def matched?
+      requests.any?
+    end
+
+    def add_matches(vcr_requests)
+      vcr_requests.each do |vcr_request|
+        next if self.requests.any? && !self.ignore_duplicates
+        next unless request_body_includes(vcr_request)
+        next unless vcr_request.method.downcase == self.method.downcase
+        matched = false
+        if self.path
+          matched = vcr_request.path == self.path
+        elsif vcr_request.path.starts_with?(self.starts_with)
+          matched = true
+          self.path = vcr_request.path
+        end
+        requests << vcr_request if matched
+      end
```

Comment:
You could consider breaking this up into a series of selects pairing down the list in a sequence of steps. This could make any future debugging of this earlier as you can check the set of requests after each select to figure out where a particular request was removed from the selection.